cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -231,32 +277,66 @@ private void writeClusterStatus(ProgramRunId programRunId, ProgramRunClusterStat
         if (artifactIdString != null) {
           artifactId = GSON.fromJson(artifactIdString, ArtifactId.class);
         }
-        String userArgumentsString = properties.get(ProgramOptionConstants.USER_OVERRIDES);
-        String systemArgumentsString = properties.get(ProgramOptionConstants.SYSTEM_OVERRIDES);
-        if (userArgumentsString == null || systemArgumentsString == null) {
-          LOG.warn("Ignore program provisioning notification for program {} without {} arguments",
-                   programRunId, (userArgumentsString == null) ? "user" : "system", notification);
+        appMetadataStore.recordProgramProvisioning(programRunId, programOptions.getUserArguments().asMap(),
+                                                   programOptions.getArguments().asMap(), messageIdBytes,
+                                                   artifactId);
+
+        // if this is a preview run or a program within a workflow, we don't actually need to provision a cluster
+        if (isInWorkflow || skipProvisioning) {
           return;
         }
-        Map<String, String> userArguments = GSON.fromJson(userArgumentsString, STRING_STRING_MAP);
-        Map<String, String> systemArguments = GSON.fromJson(systemArgumentsString, STRING_STRING_MAP);
-        appMetadataStore.recordProgramProvisioning(programRunId, userArguments, systemArguments, messageIdBytes,
-                                                   artifactId);
+
+        ProvisionRequest provisionRequest = new ProvisionRequest(programRunId, programOptions, programDescriptor,
+                                                                 userId);
+        provisioningService.provision(provisionRequest);
```

Comment:
Using provisioingService, which it uses ProvisionerStore, which will start its own tx for each message? I think it should be using the same transaction as the message subscriber in order to have messageId and state changes persisted in the same TX?

Just like we use `AppMetadataStore` directly instead of using `Store` for state updates.

Reply:
Hm right, the initial write to store that a provisioning/deprovisioning operation is happening should be in the same tx, but subsequent writes will need to be in their own tx, as they'll take place minutes later. I can change it to be like that.

Reply:
actually, writing the cluster state in the same TX as the message subscriber is problematic due to the fact that I'm writing state in a separate thread, since we expect that thread to run for minutes. If the cluster state write is done in the same tx as the subscriber, it likely will conflict with any writes done in the thread.